README.md
[Pallets](https://mods.factorio.com/mod/pallets) updated for Factorio 1.0.

A Factorio mod that allows you to load items onto pallets, and later unload them.
This makes it possible to greatly increase the throughput of a factory using belts.

Each pallet holds a single stack of items, and does not itself stack, so only belt and inserter throughput will change.

Loading and unloading pallets is very similar to filling and emptying barrels. It is done in an assembling machine. Recipes for loading and unloading are automatically created, so the mod should be compatible with other mods that introduce new items.

Loaded pallets themselves do not stack, so throughput of trains does not change when using pallets. But loading and unloading trains can be significantly faster.

Using pallets requires research to unlock.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.2.2
Date: 2020-09-14

  Bugfixes:
    - Ignore items where .stack_size is not a number.

---------------------------------------------------------------------------------------------------
Version: 0.2.1
Date: 2020-09-13

  Bugfixes:
    - Fix references to mod name to match new mod name.

---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 2020-09-13

  Changes:
    - Updated to 1.0
    - Reformatted changelog
    - Palleting research now depends on plastic bar and steel plate
    - Palleting recipes now unlock when the item is able to be created, rather than all at once.

---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 2018-01-08

  Features:
    - Make "Limit pallets to a single stack" configurable
    - Change default to 50 items per pallet

---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2018-01-07

  Major Features:
    - Initial release


render_icon.sh
#!/bin/bash
# -------------------------------------------------------------------------
#
#  renders a POV-Ray scene in different variants
#  and combines them to an icon with transparency
#
#   Usage:
#
#     render_icon.sh Scene Image Render_Options
#
#   where 'Scene' is the name of the scene
#   and 'Image the prefix for the images to generate
#
#   requires ImageMagick tools
#
# -------------------------------------------------------------------------

SCENE="$1"
IMAGE="$2"
TYPE="$3"
OPTIONS="$4"

tempdir=$(mktemp -d)
trap "rm -rf $tempdir" EXIT

thumbnail_size="+H144 +W144"
technology_size="+H128 +W128"
icon_size="+H64 +W64"

set -x
set -e

case "$TYPE" in
  thumbnail)
    povray +I${SCENE} $OPTIONS $thumbnail_size -o${tempdir}/alpha.png DECLARE=Variant=1 +ua
    povray +I${SCENE} $OPTIONS $thumbnail_size -o${tempdir}/bkg.png DECLARE=Variant=2
    povray +I${SCENE} $OPTIONS $thumbnail_size -o${tempdir}/shadow.png DECLARE=Variant=3

    shadow="( ${tempdir}/shadow.png -colorspace gray -alpha off -negate )"

    background="( ${tempdir}/bkg.png $shadow -compose copy-opacity -composite )"

    convert \( $background ${tempdir}/alpha.png -compose over -composite \) ${tempdir}/image.png

    pngcrush -rem alla -brute ${tempdir}/image.png ${IMAGE}
    ;;
  technology)
    povray +I${SCENE} $OPTIONS $technology_size -o${tempdir}/alpha.png DECLARE=Variant=1 +ua
    povray +I${SCENE} $OPTIONS $technology_size -o${tempdir}/bkg.png DECLARE=Variant=2
    povray +I${SCENE} $OPTIONS $technology_size -o${tempdir}/shadow.png DECLARE=Variant=3

    shadow="( ${tempdir}/shadow.png -colorspace gray -alpha off -negate )"

    background="( ${tempdir}/bkg.png $shadow -compose copy-opacity -composite )"

    convert \( $background ${tempdir}/alpha.png -compose over -composite \) ${tempdir}/image.png

    pngcrush -rem alla -brute ${tempdir}/image.png ${IMAGE}
    ;;
  entity)
    povray +I${SCENE} $OPTIONS -o${tempdir}/alpha.png DECLARE=Variant=4 +ua
    povray +I${SCENE} $OPTIONS -o${tempdir}/bkg.png DECLARE=Variant=5
    povray +I${SCENE} $OPTIONS -o${tempdir}/shadow.png DECLARE=Variant=6

    shadow="( ${tempdir}/shadow.png -colorspace gray -alpha off -negate )"

    convert -verbose \( ${tempdir}/bkg.png $shadow -compose copy-opacity -composite \) ${tempdir}/shadow2.png

    pngcrush -rem alla -brute ${tempdir}/shadow2.png ${IMAGE}-shadow.png
    pngcrush -rem alla -brute ${tempdir}/alpha.png   ${IMAGE}
    ;;
  *) # icon
    povray +I${SCENE} $OPTIONS $icon_size -o${tempdir}/image.png DECLARE=Variant=7 +ua
    pngcrush -rem alla -brute ${tempdir}/image.png ${IMAGE}
    ;;
esac



data-updates.lua
-- This auto-generates pallet items and fill/empty recipes for every item
-- Based on auto barreling from __base__/data-updates.lua


-- Item icon masks
local pallet_base_mask = "__pallets__/graphics/icons/pallet-base.png"
-- Recipe icon masks
local pallet_load_base_mask = "__pallets__/graphics/icons/pallet-load-base.png"
local pallet_unload_base_mask = "__pallets__/graphics/icons/pallet-unload-base.png"
local empty_pallet_item

-- Number of items that fit on a pallet
local items_per_pallet = settings.startup["pallet-stack-size"].value
local limit_items_per_pallet_to_stack_size = settings.startup["pallet-limit-to-stack"].value
-- Allow barrels on pallets?
local allow_barrels_on_pallets = false
-- Allow empty/loaded pallets onto pallets?
local allow_empty_pallets_on_pallets = true
local allow_pallets_on_pallets = false
local allow_entities_on_pallets = true
local allow_equipment_on_pallets = false
-- Should the stack size of pallets be >1?
local allow_pallet_stacks = false

-- Crafting energy per pallet load recipe
local energy_per_load = 0.2
-- Crafting energy per pallet unload recipe
local energy_per_unload = 0.2
-- If the load/unload recipes effect production statistics
local hide_palleting_from_production_stats = true
-- If the load/unloaf recipes should be included in the list of valid recipes things can use when calculating raw materials
local allow_palleting_decomposition = false

local function get_item(name)
  local items = data.raw["item"]
  if items then
    return items[name]
  end
  return nil
end

local function get_technology(name)
  local technologies = data.raw["technology"]
  if technologies then
    return technologies[name]
  end
  return nil
end

local function contains(table,element)
  if table then
    for _, value in pairs(table) do
      if value == element then
        return true
      end
    end
  end
  return false
end

local function scale_and_shift_icons(new_icons, icons, scale, shift)
  -- TODO: shifts are broken: they don't scale with the icon
  for i, icon in pairs(icons) do
    local new_icon = table.deepcopy(icon)
    
    new_icon.scale = scale * (icon.scale or 1)
    
    if icon.shift then
      new_icon.shift = {
        icon.shift[1] * scale + shift[1],
        icon.shift[2] * scale + shift[2]
      }
    else
      new_icon.shift = shift
    end
    
    new_icons[i + 1] = new_icon
  end
end

-- Generates the icons definition for a loaded pallet item
local function generate_pallet_item_icons(item)
  local new_icons = {
      {
        icon = pallet_base_mask
      }
  }
  scale_and_shift_icons(
    new_icons,
    item.icons or {
      {
        icon = item.icon
      }
    },
    0.75,
    {0, -1} 
  )
  return new_icons
end

local function generate_load_pallet_icons(item)
  local new_icons = {
      {
        icon = pallet_load_base_mask
      }
  }
  scale_and_shift_icons(
    new_icons,
    item.icons or {
      {
        icon = item.icon
      }
    },
    0.5,
    {-8, -8} 
  )
  return new_icons
end

local function generate_unload_pallet_icons(item)
  local new_icons = {
      {
        icon = pallet_unload_base_mask
      }
  }
  scale_and_shift_icons(
    new_icons,
    item.icons or {
      {
        icon = item.icon
      }
    },
    0.5,
    {8, 8} 
  )
  return new_icons
end


local function get_localised_item_name(item)
  -- For items, we should look in item-name.x
  -- For entities, look at entity-name.x
  -- This is a bit hacky, but I don't know where factorio gets the name of items that are also entities and equipment, but which don't have an item-name entry in the local
  if item.localised_name then
    return item.localised_name
  elseif item.place_result then
    return {"entity-name." .. item.name}
  elseif item.placed_as_equipment_result then
    return {"equipment-name." .. item.name}
  else
    return {"item-name." .. item.name}
  end
end

-- Generates a pallet item for stacks of the 
local function create_pallet_item(item)
  local result =
  {
    type = "item",
    name = item.name .. "-pallet",
    localised_name = {"item-name.loaded-pallet", get_localised_item_name(item)},
    icons = generate_pallet_item_icons(item),
    icon_size = 32,
    flags = {"goes-to-main-inventory"},
    subgroup = "load-pallet",
    order = "b[" .. item.name .. "-pallet]",
    stack_size = (allow_pallet_stacks and empty_pallet_item.stack_size or 1)
  }

  if contains(item.flags,"hidden") then
    table.insert(result.flags, "hidden")
  end

  data:extend({result})
  return result
end

local function get_or_create_pallet_item(item)
  local existing_item = get_item(item.name .. "-pallet")
  if existing_item then
    return existing_item
  end
  return create_pallet_item(item)
end


local function get_items_per_pallet(item)
  local n = items_per_pallet
  if limit_items_per_pallet_to_stack_size and n > item.stack_size then
    n = item.stack_size
  end
  return n
end

local function create_load_pallet_recipe(item)
  local recipe =
  {
    type = "recipe",
    name = "load-" .. item.name .. "-pallet",
    localised_name = {"recipe-name.load-pallet", get_localised_item_name(item)},
    category = "advanced-crafting",
    energy_required = energy_per_load,
    subgroup = "load-pallet",
    order = "b[load-" .. item.name .. "-pallet]",
    enabled = false,
    icons = generate_load_pallet_icons(item),
    icon_size = 32,
    ingredients =
    {
      {type = "item", name = item.name, amount = get_items_per_pallet(item)},
      {type = "item", name = empty_pallet_item.name, amount = 1},
    },
    results=
    {
      {type = "item", name = item.name .. "-pallet", amount = 1}
    },
    hide_from_stats = hide_palleting_from_production_stats,
    allow_decomposition = allow_palleting_decomposition
  }

  -- special case for pallet of pallets
  if item.name == empty_pallet_item.name then
    recipe.ingredients = {
      {type = "item", name = item.name, amount = get_items_per_pallet(item) + 1}
    }
  end

  data:extend({recipe})
  return recipe
end

local function create_unload_pallet_recipe(item)
  local recipe =
  {
    type = "recipe",
    name = "unload-" .. item.name .. "-pallet",
    localised_name = {"recipe-name.unload-pallet", get_localised_item_name(item)},
    category = "advanced-crafting",
    energy_required = energy_per_unload,
    subgroup = "unload-pallet",
    order = "c[unload-" .. item.name .. "-pallet]",
    enabled = false,
    icons = generate_unload_pallet_icons(item),
    icon_size = 32,
    ingredients =
    {
      {type = "item", name = item.name .. "-pallet", amount = 1}
    },
    results=
    {
      {type = "item", name = item.name, amount = get_items_per_pallet(item)},
      {type = "item", name = empty_pallet_item.name, amount = 1}
    },
    hide_from_stats = hide_palleting_from_production_stats,
    allow_decomposition = allow_palleting_decomposition
  }

  -- special case for pallet of pallets
  if item.name == empty_pallet_item.name then
    recipe.results = {
      {type = "item", name = item.name, amount = get_items_per_pallet(item) + 1}
    }
  end

  data:extend({recipe})
  return recipe
end

local function get_recipes_for_pallet(item)
  local recipes = data.raw["recipe"]
  if recipes then
    return recipes["load-" .. item.name .. "-pallet"], recipes["unload-" .. item.name .. "-pallet"]
  end
  return nil
end

local function get_or_create_pallet_recipes(item)
  local load_recipe, unload_recipe = get_recipes_for_pallet(item)

  if not load_recipe then
    load_recipe = create_load_pallet_recipe(item)
  end
  if not unload_recipe then
    unload_recipe = create_unload_pallet_recipe(item)
  end

  return load_recipe, unload_recipe
end


-- Adds the provided pallet recipe and load/empty recipes to the technology as recipe unlocks if they don't already exist
local function add_pallet_to_technology(load_recipe, unload_recipe, technology)
  local unlock_key = "unlock-recipe"
  local effects = technology.effects

  if not effects then
    technology.effects = {}
    effects = technology.effects
  end

  local add_load_recipe = true
  local add_unload_recipe = true

  for k,v in pairs(effects) do
    if k == unlock_key then
      local recipe = v.recipe
      if recipe == load_recipe.name then
        add_load_recipe = false
      elseif recipe == unload_recipe.name then
        add_unload_recipe = false
      end
    end
  end

  if add_load_recipe then
    table.insert(effects, {type = unlock_key, recipe = load_recipe.name})
  end
  if add_unload_recipe then
    table.insert(effects, {type = unlock_key, recipe = unload_recipe.name})
  end
end

local function allow_on_pallet(item)
  if contains(item.flags,"hidden") then
    return false
  elseif (item.auto_pallet == nil or item.auto_pallet) and (item.icon or item.icons) then
    if item.subgroup == "fill-barrel" or item.name == "empty-barrel" then
      return allow_barrels_on_pallets
    elseif item.name == "empty-pallet" then
      return allow_empty_pallets_on_pallets
    elseif item.subgroup == "load-pallet" then
      return allow_pallets_on_pallets
    elseif item.place_result then
      return allow_entities_on_pallets
    elseif item.placed_as_equipment_result then
      return allow_equipment_on_pallets
    elseif limit_items_per_pallet_to_stack_size and item.stack_size <= 1 then
      return false -- no point in stacking a single item
    else
      return true
    end
  end
  return false
end

local function process_items(items, technology)
  if not items then
    log("Auto pallet generation is disabled.")
  end

  for name,item in pairs(items) do
    if allow_on_pallet(item) then
      -- check if a pallet item already exists for this item if not - create one
      local pallet_item = get_or_create_pallet_item(item)

      -- check if the item has a palleting recipe if not - create one
      local load_recipe, unload_recipe = get_or_create_pallet_recipes(item)

      -- check if the loading/unloading recipe exists in the unlock list of the technology if not - add it
      add_pallet_to_technology(load_recipe, unload_recipe, technology)
    end
  end
end

empty_pallet_item = get_item("empty-pallet")
process_items(data.raw["item"], get_technology("pallets"))



data.lua
-- modules

local HighlyDerivative = require('__HighlyDerivative__/library')
local rusty_locale = require("__rusty-locale__.locale")
local rusty_icons = require("__rusty-locale__.icons")
local rusty_prototypes = require("__rusty-locale__.prototypes")
local ornodes = require("__OR-Nodes__/library").init()

-- external functions

local locale_of = rusty_locale.of
local icons_of = rusty_icons.of
local depend_on_all_recipe_ingredients = ornodes.depend_on_all_recipe_ingredients
local depend_on_item = ornodes.depend_on_item
local derive_name = HighlyDerivative.derive_name

-- "Constants"

local MOD_NAME = "Pallets-updated"
local PREFIX = MOD_NAME .. "-"
local MOD_DIRECTORY = "__" .. MOD_NAME .. "__/"
local GRAPHICS_DIRECTORY = MOD_DIRECTORY .. "graphics/"

local ITEMS_WITH_INDIVIDUAL_STATE = {
  ['item-with-entity-data'] = true,
  ['item-with-label'] = true,
}

local PALLET_ICON_BASE = {
  {
    icon = GRAPHICS_DIRECTORY .. "icons/empty-pallet.png",
    icon_size = 64,
    scale = 0.5
  }
}
local PALLET_ICON_OFFSET = {
  scale = 0.75,
  shift = {0, -1}
}

local LOAD_PALLET_ICON_BASE = PALLET_ICON_BASE
local LOAD_PALLET_ICON_OFFSET = PALLET_ICON_OFFSET

local UNLOAD_PALLET_ICON_BASE = {
  {
    icon = "__core__/graphics/empty.png",
    icon_size = 1,
    scale = 32,
  },
  {
    icon = GRAPHICS_DIRECTORY .. "icons/empty-pallet.png",
    icon_size = 64,
    scale = 0.25,
    shift = {8, 0}
  }
}
local UNLOAD_PALLET_ICON_OFFSET = {
  scale = 0.5,
  shift = {-8, 0}
}


local EMPTY_PALLET_ITEM_NAME = "empty-pallet"

-- static items
do
  local new_things = {}

  new_things[#new_things+1] = {
    type = "item-group",
    name = "pallets",
    order = "ca",
    icon = GRAPHICS_DIRECTORY .. "technology/pallets.png",
    icon_size = 128,
  }

  new_things[#new_things+1] = {
    type = "item-subgroup",
    name = "load-pallet",
    group = "pallets",
    order = "d"
  }

  new_things[#new_things+1] = {
    type = "item-subgroup",
    name = "unload-pallet",
    group = "pallets",
    order = "e"
  }

  new_things[#new_things+1] = {
    type = "item",
    name = EMPTY_PALLET_ITEM_NAME,
    icons = PALLET_ICON_BASE,
    subgroup = "intermediate-product",
    order = "d[empty-pallet]",
    stack_size = 10
  }

  new_things[#new_things+1] = {
    type = "recipe",
    name = EMPTY_PALLET_ITEM_NAME,
    category = "crafting",
    energy_required = 1,
    subgroup = "intermediate-product",
    enabled = false,
    ingredients =
    {
      {type="item", name="steel-plate", amount=1},
      {type="item", name="plastic-bar", amount=1},
    },
    result = EMPTY_PALLET_ITEM_NAME,
  }

  local pallet_prerequisites = {"logistics-2"}

  new_things[#new_things+1] = {
    type = "technology",
    name = "pallets",
    icon = GRAPHICS_DIRECTORY .. "technology/pallets.png",
    icon_size = 128,
    effects = {
      {
        type = "unlock-recipe",
        recipe = "empty-pallet"
      }
    },
    prerequisites = pallet_prerequisites,
    unit = {
      count = 150,
      ingredients = {
        {"automation-science-pack", 1},
        {"logistic-science-pack", 1}
      },
      time = 30
    },
    order = "c-o-a",
  }

  local steel_plate_technology = depend_on_item('steel-plate')
  log(serpent.line(steel_plate_technology))
  steel_plate_technology = steel_plate_technology[1]
  if steel_plate_technology then
    pallet_prerequisites[#pallet_prerequisites+1] = steel_plate_technology
  end

  local plastic_bar_technology = depend_on_item('plastic-bar')
  log(serpent.line(plastic_bar_technology))
  plastic_bar_technology = plastic_bar_technology[1]
  if plastic_bar_technology then
    pallet_prerequisites[#pallet_prerequisites+1] = plastic_bar_technology
  end

  data:extend(new_things)
  for i = 1,#new_things do
    -- FIXME shouldn't be necessary!
    HighlyDerivative.index(new_things[i])
  end
end

-- functions

local function is_pallet(item_name)
  return item_name:sub(1,PREFIX:len()) == PREFIX
end

local function make_pallet(new_things, item, item_name, _)
  if item.allow_on_pallet == false then return end

  local stack_size = item.stack_size or 1
  if type(stack_size) ~= 'number' then return end
  if stack_size == 1 then return end
  if stack_size > 0xffff then return end

  if is_pallet(item_name) then return end

  local names = locale_of(item)
  local icons = icons_of(item)

  local loaded_pallet_name = item_name .. "-pallet"
  if loaded_pallet_name:len() > 200 then
    loaded_pallet_name = derive_name(PREFIX, 'pallet', item_name)
  end

  local loaded_pallet = {
    type = "item",
    name = loaded_pallet_name,
    localised_name = {"item-name.loaded-pallet", names.name},
    icons = util.combine_icons(PALLET_ICON_BASE, icons, PALLET_ICON_OFFSET),
    subgroup = "load-pallet",
    order = item_name,
    stack_size = 1
  }

  local load_pallet_recipe_name = 'load-' .. item_name .. '-pallet'
  if load_pallet_recipe_name:len() > 200 then
    load_pallet_recipe_name = derive_name(PREFIX, 'load-pallet', item_name)
  end

  local load_pallet_recipe = {
    type = "recipe",
    name = load_pallet_recipe_name,
    localised_name = {"recipe-name.load-pallet", names.name},
    category = "advanced-crafting",
    energy_required = 0.2,
    subgroup = "load-pallet",
    order = item_name,
    enabled = false,
    icons = util.combine_icons(LOAD_PALLET_ICON_BASE, icons, LOAD_PALLET_ICON_OFFSET),
    ingredients =
    {
      {type = "item", name = item_name, amount = stack_size},
      {type = "item", name = EMPTY_PALLET_ITEM_NAME, amount = 1},
    },
    results=
    {
      {type = "item", name = loaded_pallet_name, amount = 1}
    },
    hide_from_stats = true,
    allow_decomposition = false
  }

  local unload_pallet_recipe_name = 'unload-' .. item_name .. '-pallet'
  if unload_pallet_recipe_name:len() > 200 then
    unload_pallet_recipe_name = derive_name(PREFIX, 'unload-pallet', item_name)
  end

  local unload_pallet_recipe =
  {
    type = "recipe",
    name = unload_pallet_recipe_name,
    localised_name = {"recipe-name.unload-pallet", names.name},
    category = "advanced-crafting",
    energy_required = 0.2,
    subgroup = "unload-pallet",
    order = item_name,
    enabled = false,
    icons = util.combine_icons(UNLOAD_PALLET_ICON_BASE, icons, UNLOAD_PALLET_ICON_OFFSET),
    ingredients =
    {
      {type = "item", name = loaded_pallet_name, amount = 1}
    },
    results=
    {
      {type = "item", name = item_name, amount = stack_size},
      {type = "item", name = EMPTY_PALLET_ITEM_NAME, amount = 1}
    },
    hide_from_stats = true,
    allow_decomposition = false
  }

  -- special case for pallet of pallets
  if item_name == EMPTY_PALLET_ITEM_NAME then
    local temp = {
      {type = "item", name = item_name, amount = stack_size + 1 }
    }
    load_pallet_recipe.ingredients = temp
    unload_pallet_recipe.results = temp
  end

  local technology_name = depend_on_all_recipe_ingredients(load_pallet_recipe, true)
  if not technology_name then
    HighlyDerivative.index()
    technology_name = depend_on_all_recipe_ingredients(load_pallet_recipe, true)
  end
  if technology_name then
    technology_name = technology_name[1]
    if not technology_name then
      -- shouldn't happen! given that we use an ingredient that is unlocked by our own technology, a technology should always be found.
      technology_name = 'pallets'
    end
  else
    technology_name = 'pallets'
  end

  local technology = data.raw.technology[technology_name]

  local function add_effects(technology_data)
    local effects = technology_data.effects
    if not effects then
      effects = {}
      technology_data.effects = effects
    end
    effects[#effects+1] = {
      type = "unlock-recipe",
      recipe = load_pallet_recipe_name
    }
    effects[#effects+1] = {
      type = "unlock-recipe",
      recipe = unload_pallet_recipe_name
    }
  end

  local normal = technology.normal
  local expensive = technology.expensive

  if normal or expensive then
    if normal then add_effects(normal) end
    if expensive then add_effects(expensive) end
  else
    add_effects(technology)
  end

  local flags = item.flags
  if flags then
    for i = 1,#flags do
      if flags[i] == "hidden" then
        loaded_pallet.flags = { "hidden" }
        load_pallet_recipe.flags = { "hidden" }
        unload_pallet_recipe.flags = { "hidden" }
      end
    end
  end

  new_things[#new_things+1] = loaded_pallet
  new_things[#new_things+1] = load_pallet_recipe
  new_things[#new_things+1] = unload_pallet_recipe
end

HighlyDerivative.register_derivation('item', make_pallet)

local function register_descendants(descendants)
  for prototype_type, value in pairs(descendants) do
    if not ITEMS_WITH_INDIVIDUAL_STATE[prototype_type] then
      HighlyDerivative.register_derivation(prototype_type, make_pallet)
      register_descendants(value)
    end
  end
end

register_descendants(rusty_prototypes.descendants('item'))